### PR TITLE
Feature/admin level map

### LINKED ISSUE #1786: New project flow: Read uploaded geojsons and show properties of individual geo area on click
>NOTE: No screenshot at the moment

## Changes

* Read geojson in the client and show temporary admin level on the map on the side
* The temporary map is a new component and replaces current map
* When a user hovers on a geo area, list all the properties available in the geo json
* Highlight what is being used in admin-level setup form on the side
* Refer to http://www.mapshaper.com/

### PR DESCRIPTION
**Pull request by samshara**
Fixes #1786
Depends on https://github.com/the-deep/server/pull/929

## Changes

* Show map on the popup
* Reflect map change when tab change

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

## This PR contains valid:

- [x] permission checks
- [] translations

### CODE REVIEW

File: src/newViews/ProjectEdit/GeoAreas/RegionCard/AddAdminLevel/index.tsx [hunk 1]
```diff
@@ -40,8 +41,11 @@ function AddAdminLevel(props: Props) {
         activeRegion,
         adminLevels: adminLevelFromProps,
         isPublished,
+        onActiveAdminLevelChange,
     } = props;
 
+    useEffect(() => () => onActiveAdminLevelChange(undefined), [onActiveAdminLevelChange]);
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/RegionCard/AddAdminLevel/index.tsx]: What does this useEffect do? We need a comment. Also the usage doesn't look right

File: src/newViews/ProjectEdit/GeoAreas/RegionCard/AddAdminLevel/index.tsx [hunk 2]
```diff
@@ -79,21 +84,24 @@ function AddAdminLevel(props: Props) {
             if (response.results.length > 0) {
                 const adminLevelsWithClientId = response.results.map(al => ({
                     ...al,
-                    clientId: randomString(),
+                    clientId: al.id.toString(),
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/RegionCard/AddAdminLevel/index.tsx]: randomString should have also worked. Why change it?

File: src/newViews/ProjectEdit/GeoAreas/index.tsx [hunk 3]
```diff
@@ -29,6 +31,10 @@ interface Props {
     activeProject: number;
 }
 
+const isAdminLevelGeoArea = (
+    value?: Partial<AdminLevelGeoArea>,
+): value is AdminLevelGeoArea => !!(value as AdminLevelGeoArea)?.id;
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/index.tsx]: Prefer function over arrow function. (arrow functions are anonymous and is a bit problematic during debugging)

File: src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+        'fill-opacity': ['case',
+            ['boolean', ['feature-state', 'hovered'], false],
+            0.7,
+            0.5,
+        ],
+        'fill-color': ['case',
+            ['boolean', ['feature-state', 'selected'], false],
+            '#1a3ed0',
+            '#a9bedc',
+        ],
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx]: do we need to have a selected state?

File: src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+import { useRequest, useLazyRequest } from '#utils/request';
+import _ts from '#ts';
+
+import styles from './styles.scss';
+
+const lineWidthScaleFactor = 2;
+const sourceOptions: AnySourceData = {
+    type: 'geojson',
+    promoteId: 'pk',
+};
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx]: Let's comment about this promoteId. Plus we have to make sure this always works

File: src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+            ['boolean', ['feature-state', 'selected'], false],
+            '#1a3ed0',
+            '#a9bedc',
+        ],
+    },
+};
+
+const mapOptions: Partial<MapboxOptions> = {
+    zoom: 2,
+    center: [50, 10],
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx]: The zoom and center pair looks arbitrary

File: src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+            }
+        },
+        failureHeader: _ts('geoAreas', 'title'),
+    });
+
+    useEffect(() => {
+        if (adminLevel) {
+            getGeoArea();
+        }
+    }, [adminLevel, getGeoArea]);
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx]: Why not just make getGeoArea a non lazy request?

File: src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+        type: 'line',
+        paint: {
+            'line-color': '#1a3ed0',
+            'line-width': ['case',
+                ['boolean', ['feature-state', 'hovered'], false],
+                thickness + 2,
+                thickness,
+            ],
+        },
+    }), [thickness]);
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx]: I think we can merge these

File: src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+
+    const handleAdminLevelChange = useCallback((id: AdminLevelGeoArea['id']) => {
+        const selected = adminLevelGeoAreas?.results.find(v => v.id === id);
+        setSelectedAdminLevel(selected);
+        getGeoJson(id);
+        getBounds(id);
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx]: Why not make getGeoJson and getBounds non-lazy and use selected as their parameter?

File: src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+    }, [getBounds, getGeoJson, adminLevelGeoAreas]);
+
+    const handleMouseEnter = useCallback((feature: MapboxGeoJSONFeature, lngLat: LngLat) => {
+        if (feature.properties) {
+            const properties = Object.entries(feature.properties)
+                .map(([key, value]) => ({ key, value }));
+            setHoverFeatureProperties(properties);
+        }
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx]: If feature.properties is not defined, we should still clear `hover features properties`

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/newViews/ProjectEdit/GeoAreas/RegionCard/AddAdminLevel/index.tsx (hunk 2 above)
Comment (samshara) [reply src/newViews/ProjectEdit/GeoAreas/RegionCard/AddAdminLevel/index.tsx]: @tnagorra to sync active tab after save

File: src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx (hunk 6 above)
Comment (samshara) [reply src/newViews/ProjectEdit/GeoAreas/RegionMap/index.tsx]: Original RegionMap had these settings

File: src/newViews/ProjectEdit/GeoAreas/index.tsx [hunk 11]
```diff
@@ -29,6 +31,10 @@ interface Props {
     activeProject: number;
 }
 
+function isAdminLevelGeoArea(value?: Partial<AdminLevelGeoArea>): value is AdminLevelGeoArea {
+    return !!(value as AdminLevelGeoArea)?.id;
+}
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/index.tsx]: What would be the different between AdminLevelGeoArea and it's Partial?
Why do we have a switch between RegionMap and RegionMapList?

Is it to show temporary map when a RegionMap is being added?

File: src/newViews/ProjectEdit/GeoAreas/RegionCard/AddAdminLevel/index.tsx [hunk 12]
```diff
@@ -111,14 +126,23 @@ function AddAdminLevel(props: Props) {
...
 
     const handlePublishGeoArea = useCallback(() => { // TODO add this later
     }, []);
 
+    const handleActiveTabChange = useCallback((id: string) => {
+        setActiveTab(id);
+        const activeAdminLevel = adminLevels.filter(v => v.id)
+            .find(v => v.clientId === id);
+        if (activeAdminLevel) {
+            onActiveAdminLevelChange(activeAdminLevel as AdminLevelGeoArea);
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/RegionCard/AddAdminLevel/index.tsx]: Why do we need to cast here?

File: src/newViews/ProjectEdit/GeoAreas/RegionCard/AddAdminLevel/index.tsx [hunk 13]
```diff
@@ -40,8 +41,19 @@ function AddAdminLevel(props: Props) {
...
 
+    /* FIXME when the parent component (ExpandableContainer) closes, this component is
+     *  unmounted. Then there is no way to remove activeAdminLevel. So we need to clear
+     *  the activeAdminLevel. Later we can add functionality to Expandable Container to
+     * take function which is called when the container is closed. Then we can remove
+     * the following code.
+     */
+    useEffect(() =>
+        () => onActiveAdminLevelChange(undefined),
+    [onActiveAdminLevelChange]);
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/RegionCard/AddAdminLevel/index.tsx]: Why do we need to clear this value? Looks like what we need is a controlled component

File: src/newViews/ProjectEdit/GeoAreas/index.tsx [hunk 14]
```diff
@@ -55,18 +62,28 @@ function GeoAreas(props: Props) {
...
-                    projectId={activeProject}
-                    regions={regions}
-                    onRegionAdd={regionsGetTrigger}
-                />
+                { activeAdminLevel ? (
+                    <AdminLevelMap
+                        className={styles.map}
+                        adminLevel={activeAdminLevel}
+                        title={activeAdminLevel.title}
+                        onFreshAdminLevel={setActiveAdminLevel}
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/index.tsx]: Is `onFreshAdminLevel` correct?

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/newViews/ProjectEdit/GeoAreas/AdminLevelMap/index.tsx [hunk 15]
```diff
@@ -0,0 +1,259 @@
+import React, { useMemo, useState, useCallback } from 'react';
```
Comment (tnagorra) [inline src/newViews/ProjectEdit/GeoAreas/AdminLevelMap/index.tsx]: This feels redundant.
If we are not immediately showing preview of uploaded geoarea, we should not even require a separate map component.

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/newViews/ProjectEdit/GeoAreas/RegionCard/AddAdminLevel/index.tsx (hunk 13 above)
Comment (samshara) [reply src/newViews/ProjectEdit/GeoAreas/RegionCard/AddAdminLevel/index.tsx]: @tnagorra when the expandable container is closed we need to remove the set activeTab value

**Review by AdityaKhatri — APPROVED** [review]

**Pull request merged**